In [43]:
import numpy as np
import os, tempfile
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import statsmodels.api as sm
import tensorflow as tf
from tensorflow.keras import models, layers
from tensorflow.keras.preprocessing.image import load_img

from tensorflow.keras.models import Sequential, load_model
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.regularizers import l2
from tensorflow.keras.losses import MeanSquaredLogarithmicError
from sklearn.model_selection import cross_validate
from sklearn.dummy import DummyClassifier
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.datasets import make_regression

from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge

from sklearn.dummy import DummyRegressor
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.datasets import make_regression
from sklearn.linear_model import LinearRegression
from sklearn.dummy import DummyRegressor
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score
from sklearn.metrics import plot_confusion_matrix
from sklearn.metrics import roc_auc_score, plot_roc_curve

from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import GradientBoostingRegressor

from sklearn.neighbors import KNeighborsRegressor
from sklearn.neighbors import KNeighborsClassifier

from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score
from sklearn.metrics import plot_confusion_matrix
from sklearn.metrics import roc_auc_score, plot_roc_curve

from sklearn.tree import export_graphviz
from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import DecisionTreeRegressor

import sklearn.metrics as metrics
from scipy import stats as stats
import statsmodels.api as sm
from statsmodels.formula.api import ols
%matplotlib inline
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay
from sklearn.model_selection import train_test_split

In [44]:
pd17 = pd.read_csv("./data/2017nfldata.csv")
pd18 = pd.read_csv("./data/2018nfldata.csv")
pd19 = pd.read_csv("./data/2019nfldata.csv")
pd20 = pd.read_csv("./data/2020nfldata.csv")
pd21 = pd.read_csv("./data/2021nfldata.csv")

In [45]:
pd19.head(3)

Rk                 Player   Tm FantPos  Age   G  GS  Cmp  Att   Yds  ...  \
0   1  Christian McCaffrey*+  CAR      RB   23  16  16    0    2     0  ...   
1   2        Lamar Jackson*+  BAL      QB   22  15  15  265  401  3127  ...   
2   3        Derrick Henry *  TEN      RB   25  15  15    0    0     0  ...   

   2PM  2PP  FantPt    PPR   DKPt   FDPt    VBD  PosRank  OvRank     -9999  
0  1.0  NaN   355.0  471.2  477.2  413.2  215.0        1     1.0  McCaCh01  
1  NaN  NaN   416.0  415.7  429.7  421.7  152.0        1     2.0  JackLa00  
2  NaN  NaN   277.0  294.6  303.6  285.6  136.0        2     3.0  HenrDe00  

[3 rows x 34 columns]

In [46]:
pd20.head(3)

Rk            Player   Tm FantPos  Age   G  GS  Cmp  Att  Yds  ...  2PM  \
0   1  Derrick Henry *+  TEN      RB   26  16  16    0    0    0  ...  1.0   
1   2     Alvin Kamara*  NOR      RB   25  15  10    0    0    0  ...  NaN   
2   3      Dalvin Cook*  MIN      RB   25  14  14    0    0    0  ...  3.0   

   2PP  FantPt    PPR   DKPt   FDPt    VBD  PosRank  OvRank     -9999  
0  NaN   314.0  333.1  341.1  323.6  184.0        1     1.0  HenrDe00  
1  NaN   295.0  377.8  383.8  336.3  165.0        2     2.0  KamaAl00  
2  NaN   294.0  337.8  346.8  315.8  164.0        3     3.0  CookDa01  

[3 rows x 34 columns]

In [47]:
pd21.head(3)

Rk             Player   Tm FantPos  Age   G  GS  Cmp  Att   Yds  ...  2PM  \
0   1  Jonathan Taylor*+  IND      RB   22  17  17  0.0  0.0   0.0  ...  NaN   
1   2      Cooper Kupp*+  LAR      WR   28  17  17  0.0  1.0   0.0  ...  1.0   
2   3     Deebo Samuel*+  SFO      WR   25  16  15  1.0  2.0  24.0  ...  NaN   

   2PP  FantPt    PPR   DKPt   FDPt    VBD  PosRank  OvRank     -9999  
0  NaN   333.0  373.1  381.1  353.1  187.0        1     1.0  TaylJo02  
1  NaN   295.0  439.5  442.5  367.0  173.0        1     2.0  KuppCo00  
2  NaN   262.0  339.0  347.0  300.5  140.0        2     3.0  SamuDe00  

[3 rows x 34 columns]

In [48]:
for col in pd17.columns:
    print(col)

Rk
Player
Tm
FantPos
Age
G
GS
Cmp
Att
Yds
TD
Int
Att.1
Yds.1
Y/A
TD.1
Tgt
Rec
Yds.2
Y/R
TD.2
Fmb
FL
TD.3
2PM
2PP
FantPt
PPR
DKPt
FDPt
VBD
PosRank
OvRank
-9999


In [49]:
for col in pd18.columns:
    print(col)

Rk
Player
Tm
FantPos
Age
G
GS
Cmp
Att
Yds
TD
Int
Att.1
Yds.1
Y/A
TD.1
Tgt
Rec
Yds.2
Y/R
TD.2
Fmb
FL
TD.3
2PM
2PP
FantPt
PPR
DKPt
FDPt
VBD
PosRank
OvRank
-9999


In [50]:
pd17 = pd17.drop(columns=['Rk', 'Player', 'FantPt', 'DKPt', 'FDPt', 'VBD', 'PosRank', 'OvRank'])
pd18 = pd18.drop(columns=['Rk', 'Player', 'FantPt', 'DKPt', 'FDPt', 'VBD', 'PosRank', 'OvRank'])
pd19 = pd19.drop(columns=['Rk', 'Player', 'FantPt', 'DKPt', 'FDPt', 'VBD', 'PosRank', 'OvRank'])
pd20 = pd20.drop(columns=['Rk', 'Player', 'FantPt', 'DKPt', 'FDPt', 'VBD', 'PosRank', 'OvRank'])
pd21 = pd21.drop(columns=['Rk', 'Player', 'FantPt', 'DKPt', 'FDPt', 'VBD', 'PosRank', 'OvRank'])

In [51]:
pd17.fillna(0, inplace=True)
pd18.fillna(0, inplace=True)
pd19.fillna(0, inplace=True)
pd20.fillna(0, inplace=True)
pd21.fillna(0, inplace=True)


In [52]:
pd17.head()

Tm FantPos  Age   G  GS  Cmp  Att  Yds  TD  Int  ...  Yds.2    Y/R  TD.2  \
0  LAR      RB   23  15  15    0    0    0   0    0  ...    788  12.31     6   
1  PIT      RB   25  15  15    0    0    0   0    0  ...    655   7.71     2   
2  KAN      RB   22  16  16    0    0    0   0    0  ...    455   8.58     3   
3  NOR      RB   22  16   3    0    0    0   0    0  ...    826  10.20     5   
4  LAC      RB   24  16  16    0    0    0   0    0  ...    476   8.21     4   

   Fmb  FL  TD.3  2PM  2PP    PPR     -9999  
0  5.0   2    19  0.0  0.0  383.3  GurlTo01  
1  3.0   2    11  0.0  0.0  341.6  BellLe00  
2  1.0   1    11  0.0  0.0  295.2  HuntKa00  
3  1.0   1    14  1.0  0.0  320.4  KamaAl00  
4  1.0   0    12  0.0  0.0  288.1  GordMe00  

[5 rows x 26 columns]

In [53]:
pd18.head()

Tm FantPos  Age   G  GS  Cmp  Att   Yds  TD  Int  ...  Yds.2   Y/R  TD.2  \
0  LAR      RB   24  14  14    0    0     0   0    0  ...    580  9.83     4   
1  NYG      RB   21  16  16    0    0     0   0    0  ...    721  7.92     4   
2  CAR      RB   22  16  16    1    1    50   1    0  ...    867  8.10     6   
3  NOR      RB   23  15  13    0    0     0   0    0  ...    709  8.75     4   
4  KAN      QB   23  16  16  383  580  5097  50   12  ...      0  0.00     0   

   Fmb  FL  TD.3  2PM  2PP    PPR     -9999  
0  1.0   1    21  3.0  0.0  372.1  GurlTo01  
1  0.0   0    15  1.0  0.0  385.8  BarkSa00  
2  4.0   1    13  0.0  0.0  385.5  McCaCh01  
3  1.0   0    18  3.0  0.0  354.2  KamaAl00  
4  9.0   2     2  1.0  0.0  417.1  MahoPa00  

[5 rows x 26 columns]

In [54]:
pd17target = pd18[['PPR','-9999']]

In [55]:
pd17target.head()

PPR     -9999
0  372.1  GurlTo01
1  385.8  BarkSa00
2  385.5  McCaCh01
3  354.2  KamaAl00
4  417.1  MahoPa00

In [56]:
pd18target = pd19[['PPR','-9999']]
pd19target = pd20[['PPR','-9999']]
pd20target = pd21[['PPR','-9999']]


In [57]:
pd17target.rename(columns={'PPR':'PPRtarget'}, inplace=True)

C:\Users\Peter\anaconda3\envs\learn-env\lib\site-packages\pandas\core\frame.py:4296: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


In [58]:
pd17target.head()

PPRtarget     -9999
0      372.1  GurlTo01
1      385.8  BarkSa00
2      385.5  McCaCh01
3      354.2  KamaAl00
4      417.1  MahoPa00

In [59]:
pd18target.rename(columns={'PPR':'PPRtarget'}, inplace=True)
pd19target.rename(columns={'PPR':'PPRtarget'}, inplace=True)
pd20target.rename(columns={'PPR':'PPRtarget'}, inplace=True)

C:\Users\Peter\anaconda3\envs\learn-env\lib\site-packages\pandas\core\frame.py:4296: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


In [60]:
pd18target.head()

PPRtarget     -9999
0      471.2  McCaCh01
1      415.7  JackLa00
2      294.6  HenrDe00
3      314.8  JoneAa00
4      311.7  ElliEz00

In [61]:
pd17final = pd17.merge(pd17target,how='left', left_on='-9999', right_on='-9999')

In [62]:
pd17final.head()

Tm FantPos  Age   G  GS  Cmp  Att  Yds  TD  Int  ...    Y/R  TD.2  Fmb  \
0  LAR      RB   23  15  15    0    0    0   0    0  ...  12.31     6  5.0   
1  PIT      RB   25  15  15    0    0    0   0    0  ...   7.71     2  3.0   
2  KAN      RB   22  16  16    0    0    0   0    0  ...   8.58     3  1.0   
3  NOR      RB   22  16   3    0    0    0   0    0  ...  10.20     5  1.0   
4  LAC      RB   24  16  16    0    0    0   0    0  ...   8.21     4  1.0   

   FL  TD.3  2PM  2PP    PPR     -9999  PPRtarget  
0   2    19  0.0  0.0  383.3  GurlTo01      372.1  
1   2    11  0.0  0.0  341.6  BellLe00        NaN  
2   1    11  0.0  0.0  295.2  HuntKa00      230.2  
3   1    14  1.0  0.0  320.4  KamaAl00      354.2  
4   0    12  0.0  0.0  288.1  GordMe00      275.5  

[5 rows x 27 columns]

In [63]:
pd18final = pd18.merge(pd18target,how='left', left_on='-9999', right_on='-9999')
pd19final = pd19.merge(pd19target,how='left', left_on='-9999', right_on='-9999')
pd20final = pd20.merge(pd20target,how='left', left_on='-9999', right_on='-9999')

In [64]:
df_final = pd.DataFrame().append([pd18final,pd19final, pd20final, pd17final])

In [65]:
df_final.head()

Tm FantPos  Age   G  GS  Cmp  Att   Yds  TD  Int  ...   Y/R  TD.2  Fmb  \
0  LAR      RB   24  14  14    0    0     0   0    0  ...  9.83     4  1.0   
1  NYG      RB   21  16  16    0    0     0   0    0  ...  7.92     4  0.0   
2  CAR      RB   22  16  16    1    1    50   1    0  ...  8.10     6  4.0   
3  NOR      RB   23  15  13    0    0     0   0    0  ...  8.75     4  1.0   
4  KAN      QB   23  16  16  383  580  5097  50   12  ...  0.00     0  9.0   

   FL  TD.3  2PM  2PP    PPR     -9999  PPRtarget  
0   1    21  3.0  0.0  372.1  GurlTo01      219.4  
1   0    15  1.0  0.0  385.8  BarkSa00      244.1  
2   1    13  0.0  0.0  385.5  McCaCh01      471.2  
3   0    18  3.0  0.0  354.2  KamaAl00      248.5  
4   2     2  1.0  0.0  417.1  MahoPa00      287.0  

[5 rows x 27 columns]

In [66]:
df_final = df_final.sort_values('PPR', ascending=False)

In [67]:
df_final.head(20)

Tm FantPos  Age   G  GS  Cmp  Att   Yds  TD  Int  ...    Y/R  TD.2   Fmb  \
0   CAR      RB   23  16  16    0    2     0   0    0  ...   8.66     4   1.0   
4   KAN      QB   23  16  16  383  580  5097  50   12  ...   0.00     0   9.0   
1   BAL      QB   22  15  15  265  401  3127  36    6  ...   0.00     0   9.0   
6   BUF      QB   24  16  16  396  572  4544  37   10  ...  12.00     1   9.0   
1   NYG      RB   21  16  16    0    0     0   0    0  ...   7.92     4   0.0   
2   CAR      RB   22  16  16    1    1    50   1    0  ...   8.10     6   4.0   
0   LAR      RB   23  15  15    0    0     0   0    0  ...  12.31     6   5.0   
7   GNB      QB   37  16  16  372  526  4299  48    5  ...  -6.00     0   4.0   
8   ARI      QB   23  16  16  375  558  3971  26   12  ...   0.00     0   9.0   
1   NOR      RB   25  15  10    0    0     0   0    0  ...   9.11     5   1.0   
6   NOR      WR   26  16  15    0    0     0   0    0  ...  11.58     9   1.0   
9   KAN      QB   25  15  15  390  588  4740  38    6  ...   0.00     0   5.0   
0   LAR      RB   24  14  14    0    0     0   0    0  ...   9.83     4   1.0   
10  HOU      QB   25  16  16  382  544  4823  33    7  ...   0.00     0   8.0   
15  SEA      QB   32  16  16  384  558  4212  40   13  ...   0.00     0   7.0   
3   GNB      WR   28  14  14    0    0     0   0    0  ...  11.95    18   1.0   
20  ATL      QB   33  16  16  422  608  4924  35    7  ...   5.00     1  10.0   
3   NOR      RB   23  15  13    0    0     0   0    0  ...   8.75     4   1.0   
8   SEA      QB   29  16  16  339  553  3983  34   11  ...   0.00     0  14.0   
22  TEN      QB   32  16  16  315  481  3819  33    7  ...   0.00     0   6.0   

    FL  TD.3  2PM  2PP    PPR     -9999  PPRtarget  
0    0    19  1.0  0.0  471.2  McCaCh01       90.4  
4    2     2  1.0  0.0  417.1  MahoPa00      287.0  
1    2     7  0.0  0.0  415.7  JackLa00      332.8  
6    6     9  0.0  0.0  396.1  AlleJo02      402.6  
1    0    15  1.0  0.0  385.8  BarkSa00      244.1  
2    1    13  0.0  0.0  385.5  McCaCh01      471.2  
0    2    19  0.0  0.0  383.3  GurlTo01      372.1  
7    2     3  0.0  0.0  383.3  RodgAa00      333.3  
8    4    11  0.0  0.0  378.7  MurrKy00      300.5  
1    0    21  0.0  0.0  377.8  KamaAl00      234.7  
6    0     9  0.0  0.0  374.6  ThomMi05       83.9  
9    2     2  0.0  3.0  374.4  MahoPa00      361.7  
0    1    21  3.0  0.0  372.1  GurlTo01      219.4  
10   3     3  0.0  1.0  369.3  WatsDe00        NaN  
15   4     2  0.0  1.0  359.8  WilsRu00      242.8  
3    1    18  0.0  0.0  358.4  AdamDa01      344.3  
20   5     4  0.0  2.0  355.0  RyanMa00      267.3  
3    0    18  3.0  0.0  354.2  KamaAl00      248.5  
8    3     3  0.0  2.0  347.9  WilsRu00      299.4  
22   1     7  2.0  1.0  344.4  TannRy00      268.4  

[20 rows x 27 columns]

In [68]:
df_final = df_final[df_final['PPR'] > 0]  

In [69]:
df_final = df_final.dropna(subset = ['PPR'])

In [70]:
df_final = df_final.dropna(subset = ['PPRtarget'])

In [71]:
df_final.shape

(1711, 27)

In [72]:
df_final.head(20)

Tm FantPos  Age   G  GS  Cmp  Att   Yds  TD  Int  ...    Y/R  TD.2   Fmb  \
0   CAR      RB   23  16  16    0    2     0   0    0  ...   8.66     4   1.0   
4   KAN      QB   23  16  16  383  580  5097  50   12  ...   0.00     0   9.0   
1   BAL      QB   22  15  15  265  401  3127  36    6  ...   0.00     0   9.0   
6   BUF      QB   24  16  16  396  572  4544  37   10  ...  12.00     1   9.0   
1   NYG      RB   21  16  16    0    0     0   0    0  ...   7.92     4   0.0   
2   CAR      RB   22  16  16    1    1    50   1    0  ...   8.10     6   4.0   
0   LAR      RB   23  15  15    0    0     0   0    0  ...  12.31     6   5.0   
7   GNB      QB   37  16  16  372  526  4299  48    5  ...  -6.00     0   4.0   
8   ARI      QB   23  16  16  375  558  3971  26   12  ...   0.00     0   9.0   
1   NOR      RB   25  15  10    0    0     0   0    0  ...   9.11     5   1.0   
6   NOR      WR   26  16  15    0    0     0   0    0  ...  11.58     9   1.0   
9   KAN      QB   25  15  15  390  588  4740  38    6  ...   0.00     0   5.0   
0   LAR      RB   24  14  14    0    0     0   0    0  ...   9.83     4   1.0   
15  SEA      QB   32  16  16  384  558  4212  40   13  ...   0.00     0   7.0   
3   GNB      WR   28  14  14    0    0     0   0    0  ...  11.95    18   1.0   
20  ATL      QB   33  16  16  422  608  4924  35    7  ...   5.00     1  10.0   
3   NOR      RB   23  15  13    0    0     0   0    0  ...   8.75     4   1.0   
8   SEA      QB   29  16  16  339  553  3983  34   11  ...   0.00     0  14.0   
22  TEN      QB   32  16  16  315  481  3819  33    7  ...   0.00     0   6.0   
26  PIT      QB   36  16  16  452  675  5129  34   16  ...  -1.00     0   7.0   

    FL  TD.3  2PM  2PP    PPR     -9999  PPRtarget  
0    0    19  1.0  0.0  471.2  McCaCh01       90.4  
4    2     2  1.0  0.0  417.1  MahoPa00      287.0  
1    2     7  0.0  0.0  415.7  JackLa00      332.8  
6    6     9  0.0  0.0  396.1  AlleJo02      402.6  
1    0    15  1.0  0.0  385.8  BarkSa00      244.1  
2    1    13  0.0  0.0  385.5  McCaCh01      471.2  
0    2    19  0.0  0.0  383.3  GurlTo01      372.1  
7    2     3  0.0  0.0  383.3  RodgAa00      333.3  
8    4    11  0.0  0.0  378.7  MurrKy00      300.5  
1    0    21  0.0  0.0  377.8  KamaAl00      234.7  
6    0     9  0.0  0.0  374.6  ThomMi05       83.9  
9    2     2  0.0  3.0  374.4  MahoPa00      361.7  
0    1    21  3.0  0.0  372.1  GurlTo01      219.4  
15   4     2  0.0  1.0  359.8  WilsRu00      242.8  
3    1    18  0.0  0.0  358.4  AdamDa01      344.3  
20   5     4  0.0  2.0  355.0  RyanMa00      267.3  
3    0    18  3.0  0.0  354.2  KamaAl00      248.5  
8    3     3  0.0  2.0  347.9  WilsRu00      299.4  
22   1     7  2.0  1.0  344.4  TannRy00      268.4  
26   2     3  0.0  4.0  341.9  RoetBe00       12.7  

[20 rows x 27 columns]

In [73]:
df_final.fillna(0, inplace=True)

In [74]:
df_final = df_final.drop(columns=['-9999'])

In [75]:
df_final.isna().sum()

Tm           0
FantPos      0
Age          0
G            0
GS           0
Cmp          0
Att          0
Yds          0
TD           0
Int          0
Att.1        0
Yds.1        0
Y/A          0
TD.1         0
Tgt          0
Rec          0
Yds.2        0
Y/R          0
TD.2         0
Fmb          0
FL           0
TD.3         0
2PM          0
2PP          0
PPR          0
PPRtarget    0
dtype: int64

In [76]:
df_final.shape

(1711, 26)

In [77]:
df_final.head(20)

Tm FantPos  Age   G  GS  Cmp  Att   Yds  TD  Int  ...  Yds.2    Y/R  \
0   CAR      RB   23  16  16    0    2     0   0    0  ...   1005   8.66   
4   KAN      QB   23  16  16  383  580  5097  50   12  ...      0   0.00   
1   BAL      QB   22  15  15  265  401  3127  36    6  ...      0   0.00   
6   BUF      QB   24  16  16  396  572  4544  37   10  ...     12  12.00   
1   NYG      RB   21  16  16    0    0     0   0    0  ...    721   7.92   
2   CAR      RB   22  16  16    1    1    50   1    0  ...    867   8.10   
0   LAR      RB   23  15  15    0    0     0   0    0  ...    788  12.31   
7   GNB      QB   37  16  16  372  526  4299  48    5  ...     -6  -6.00   
8   ARI      QB   23  16  16  375  558  3971  26   12  ...      0   0.00   
1   NOR      RB   25  15  10    0    0     0   0    0  ...    756   9.11   
6   NOR      WR   26  16  15    0    0     0   0    0  ...   1725  11.58   
9   KAN      QB   25  15  15  390  588  4740  38    6  ...      0   0.00   
0   LAR      RB   24  14  14    0    0     0   0    0  ...    580   9.83   
15  SEA      QB   32  16  16  384  558  4212  40   13  ...      0   0.00   
3   GNB      WR   28  14  14    0    0     0   0    0  ...   1374  11.95   
20  ATL      QB   33  16  16  422  608  4924  35    7  ...      5   5.00   
3   NOR      RB   23  15  13    0    0     0   0    0  ...    709   8.75   
8   SEA      QB   29  16  16  339  553  3983  34   11  ...      0   0.00   
22  TEN      QB   32  16  16  315  481  3819  33    7  ...      0   0.00   
26  PIT      QB   36  16  16  452  675  5129  34   16  ...     -1  -1.00   

    TD.2   Fmb  FL  TD.3  2PM  2PP    PPR  PPRtarget  
0      4   1.0   0    19  1.0  0.0  471.2       90.4  
4      0   9.0   2     2  1.0  0.0  417.1      287.0  
1      0   9.0   2     7  0.0  0.0  415.7      332.8  
6      1   9.0   6     9  0.0  0.0  396.1      402.6  
1      4   0.0   0    15  1.0  0.0  385.8      244.1  
2      6   4.0   1    13  0.0  0.0  385.5      471.2  
0      6   5.0   2    19  0.0  0.0  383.3      372.1  
7      0   4.0   2     3  0.0  0.0  383.3      333.3  
8      0   9.0   4    11  0.0  0.0  378.7      300.5  
1      5   1.0   0    21  0.0  0.0  377.8      234.7  
6      9   1.0   0     9  0.0  0.0  374.6       83.9  
9      0   5.0   2     2  0.0  3.0  374.4      361.7  
0      4   1.0   1    21  3.0  0.0  372.1      219.4  
15     0   7.0   4     2  0.0  1.0  359.8      242.8  
3     18   1.0   1    18  0.0  0.0  358.4      344.3  
20     1  10.0   5     4  0.0  2.0  355.0      267.3  
3      4   1.0   0    18  3.0  0.0  354.2      248.5  
8      0  14.0   3     3  0.0  2.0  347.9      299.4  
22     0   6.0   1     7  2.0  1.0  344.4      268.4  
26     0   7.0   2     3  0.0  4.0  341.9       12.7  

[20 rows x 26 columns]

In [78]:
df_final.tail()

Tm FantPos  Age   G  GS  Cmp  Att  Yds  TD  Int  ...  Yds.2   Y/R  TD.2  \
611  IND      RB   29   2   0    0    0    0   0    0  ...      3   3.0     0   
592  ATL      WR   29  16   0    0    0    0   0    0  ...     12  12.0     0   
535  LAR      RB   22   4   0    0    0    0   0    0  ...      0   0.0     0   
574  OAK      QB   28   2   0    1    3    8   0    0  ...      0   0.0     0   
580  BAL      QB   24   1   0    0    0    0   0    0  ...      0   0.0     0   

     Fmb  FL  TD.3  2PM  2PP  PPR  PPRtarget  
611  1.0   1     0  0.0  0.0  0.3        0.0  
592  4.0   1     0  0.0  0.0  0.2       35.9  
535  0.0   0     0  0.0  0.0  0.1        1.9  
574  1.0   0     0  0.0  0.0  0.1       16.9  
580  0.0   0     0  0.0  0.0  0.1        9.3  

[5 rows x 26 columns]

## Correlation Between Variables in Dataset

In [79]:
trainset = X_train.join(y_train)

NameError: name 'X_train' is not defined

In [80]:
sns.heatmap(trainset.corr());

NameError: name 'trainset' is not defined

In [ ]:
trainset.corr()

In [ ]:
cr = trainset.corr()
cr = cr[cr < 1]

In [ ]:
cr.unstack().sort_values(ascending=False)[:30]

In [81]:
## sns.pairplot(trainset.head(10))

## Train Test Split

In [82]:
X = df_final.drop(columns=['PPRtarget'])
y = df_final['PPRtarget']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42) 

In [83]:
X_train.shape

(1368, 25)

In [84]:
y_train.shape

(1368,)

## Feature engineering

In order to analyze our categorical vaiables, we will convert the team each player was on, and the position that they play with one hot encoding

In [85]:
categoricals = ['Tm', 'FantPos']
X_train_dummies = pd.get_dummies(X_train[categoricals], prefix=categoricals)
X_train_dummies

Tm_2TM  Tm_3TM  Tm_ARI  Tm_ATL  Tm_BAL  Tm_BUF  Tm_CAR  Tm_CHI  Tm_CIN  \
110       0       0       0       0       0       0       1       0       0   
246       0       0       0       0       0       0       0       0       0   
294       0       0       0       0       0       0       0       0       0   
99        0       0       0       0       0       0       0       0       0   
255       0       0       0       0       0       0       0       0       0   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
306       0       0       0       0       0       0       0       1       0   
383       0       0       0       0       0       0       0       0       0   
192       1       0       0       0       0       0       0       0       0   
467       0       0       0       0       0       0       0       0       1   
313       0       0       0       0       0       0       0       0       0   

     Tm_CLE  ...  Tm_PIT  Tm_SEA  Tm_SFO  Tm_TAM  Tm_TEN  Tm_WAS  FantPos_QB  \
110       0  ...       0       0       0       0       0       0           0   
246       0  ...       0       0       0       0       0       0           0   
294       0  ...       0       0       0       0       0       0           0   
99        0  ...       0       0       0       1       0       0           0   
255       0  ...       0       0       1       0       0       0           0   
..      ...  ...     ...     ...     ...     ...     ...     ...         ...   
306       0  ...       0       0       0       0       0       0           0   
383       0  ...       0       0       0       0       0       0           0   
192       0  ...       0       0       0       0       0       0           0   
467       0  ...       0       0       0       0       0       0           0   
313       0  ...       0       0       0       0       0       0           0   

     FantPos_RB  FantPos_TE  FantPos_WR  
110           1           0           0  
246           0           1           0  
294           1           0           0  
99            1           0           0  
255           1           0           0  
..          ...         ...         ...  
306           0           0           1  
383           1           0           0  
192           1           0           0  
467           0           1           0  
313           0           0           1  

[1368 rows x 39 columns]

In [86]:
categoricals = ['Tm', 'FantPos']
X_test_dummies = pd.get_dummies(X_test[categoricals], prefix=categoricals)
X_test_dummies

Tm_2TM  Tm_ARI  Tm_ATL  Tm_BAL  Tm_BUF  Tm_CAR  Tm_CHI  Tm_CIN  Tm_CLE  \
311       0       0       0       0       0       0       0       0       0   
368       0       0       0       1       0       0       0       0       0   
439       0       0       0       0       0       0       0       0       0   
24        0       0       0       0       0       0       0       0       0   
364       0       0       0       0       0       0       0       0       0   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
502       0       0       0       0       0       0       0       0       0   
285       0       0       0       0       0       0       0       0       0   
40        0       0       0       0       0       0       0       0       0   
107       0       1       0       0       0       0       0       0       0   
382       0       0       0       0       0       0       0       0       1   

     Tm_DAL  ...  Tm_PIT  Tm_SEA  Tm_SFO  Tm_TAM  Tm_TEN  Tm_WAS  FantPos_QB  \
311       0  ...       0       0       1       0       0       0           0   
368       0  ...       0       0       0       0       0       0           0   
439       0  ...       0       0       0       0       0       0           0   
24        0  ...       0       0       0       0       0       0           0   
364       0  ...       0       0       0       0       0       0           0   
..      ...  ...     ...     ...     ...     ...     ...     ...         ...   
502       0  ...       0       0       0       1       0       0           0   
285       0  ...       0       0       0       0       0       0           0   
40        0  ...       0       0       0       0       0       0           0   
107       0  ...       0       0       0       0       0       0           1   
382       0  ...       0       0       0       0       0       0           0   

     FantPos_RB  FantPos_TE  FantPos_WR  
311           0           0           1  
368           1           0           0  
439           1           0           0  
24            0           0           1  
364           0           1           0  
..          ...         ...         ...  
502           0           1           0  
285           0           0           1  
40            1           0           0  
107           0           0           0  
382           0           1           0  

[343 rows x 38 columns]

Dropping original categorical values so that we can add dummy columns:

In [87]:
X_train.drop(['Tm', 'FantPos'], axis=1, inplace=True)
X_test.drop(['Tm', 'FantPos'], axis=1, inplace=True)

C:\Users\Peter\anaconda3\envs\learn-env\lib\site-packages\pandas\core\frame.py:4163: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


Removing data for data on players that played on multiple teams in one season:

In [88]:
X_train_dummies.drop(['Tm_2TM', 'Tm_3TM'], axis=1, inplace=True)
X_test_dummies.drop(['Tm_2TM'], axis=1, inplace=True)

In [89]:
X_train.shape

(1368, 23)

In [90]:
X_train_dummies.shape

(1368, 37)

Joining our catergorical data to our test/train sets:

In [91]:
X_train = pd.concat([X_train, X_train_dummies], axis="columns")

In [92]:
X_test = pd.concat([X_test, X_test_dummies], axis="columns")

In [93]:
X_train.shape


(1368, 60)

In [94]:
X_train.head(5)

Age   G  GS  Cmp  Att  Yds  TD  Int  Att.1  Yds.1  ...  Tm_PIT  Tm_SEA  \
110   30  15  10    0    0    0   0    0    198    680  ...       0       0   
246   29  16  11    0    0    0   0    0      0      0  ...       0       0   
294   23  12   0    0    0    0   0    0     36    191  ...       0       0   
99    24  16  16    0    0    0   0    0    234    871  ...       0       0   
255   27  16  14    0    0    0   0    0      8     30  ...       0       0   

     Tm_SFO  Tm_TAM  Tm_TEN  Tm_WAS  FantPos_QB  FantPos_RB  FantPos_TE  \
110       0       0       0       0           0           1           0   
246       0       0       0       0           0           0           1   
294       0       0       0       0           0           1           0   
99        0       1       0       0           0           1           0   
255       1       0       0       0           0           1           0   

     FantPos_WR  
110           0  
246           0  
294           0  
99            0  
255           0  

[5 rows x 60 columns]

Now that we have one hot encoded team and position, lets take a look at doing something similar with age. Since a value of 33 vs a value of 27 is not that different as an absolute value for or model, but it is significant in determining an NFL players age, ie mid career vs close to retirement for most players, I am going to divide up the ages of players into discreete bins. For the purpse of fantasy analysis, I am grouping players by 3 groups. Young players(<25), who are rookies or in their first 2 years of data in the league in most cases and establishing their value still. Mid-career players(25-29) who have established their performance and typically are not showing droppoff yet, and late career players(30+) where each additional year in the league significantly increases the chance they will retire. 

In [95]:
X_test["Age Bins"] = pd.cut(X_test['Age'], 
    [0, 24, 29, 100], 
    labels=['A', 'B', 'C'],
    right=True               
)
X_train["Age Bins"] = pd.cut(X_train['Age'], 
    [0, 24, 29, 100], 
    labels=['A', 'B', 'C'],
    right=True               
)

In [96]:
X_test.head()

Age   G  GS  Cmp  Att  Yds  TD  Int  Att.1  Yds.1  ...  Tm_SEA  Tm_SFO  \
311   24  11   4    1    1   16   0    0      0      0  ...       0       1   
368   23  16   0    0    0    0   0    0      0      0  ...       0       0   
439   25  16   4    0    0    0   0    0      6     24  ...       0       0   
24    24  16  14    0    0    0   0    0      0      0  ...       0       0   
364   31  15   4    0    0    0   0    0      0      0  ...       0       0   

     Tm_TAM  Tm_TEN  Tm_WAS  FantPos_QB  FantPos_RB  FantPos_TE  FantPos_WR  \
311       0       0       0           0           0           0           1   
368       0       0       0           0           1           0           0   
439       0       0       0           0           1           0           0   
24        0       0       0           0           0           0           1   
364       0       0       0           0           0           1           0   

     Age Bins  
311         A  
368         A  
439         B  
24          A  
364         C  

[5 rows x 61 columns]

Now that we can see that we have a dataframe that mirrors X_test but with A,B,C representing ages, lets remove the original "Age" column, and then one hot encode the age categoricals. 

In [97]:
X_test_age_dummies = pd.get_dummies(X_test["Age Bins"], prefix='age')
X_train_age_dummies = pd.get_dummies(X_train["Age Bins"], prefix='age')
X_test_age_dummies

age_A  age_B  age_C
311      1      0      0
368      1      0      0
439      0      1      0
24       1      0      0
364      0      0      1
..     ...    ...    ...
502      0      1      0
285      0      0      1
40       1      0      0
107      1      0      0
382      1      0      0

[343 rows x 3 columns]

Lets add that to our dataframe. Also can remove one of the age catergories since if a player's A or B catergory encoding is known, by default we know if they are in C or not 

In [98]:
X_test = pd.concat([X_test, X_test_age_dummies], axis="columns")
X_test.drop(['age_A'], axis=1, inplace=True)
X_test.drop(['Age Bins'], axis=1, inplace=True)
X_test.drop(['Age'], axis=1, inplace=True)

X_train = pd.concat([X_train, X_train_age_dummies], axis="columns")
X_train.drop(['age_A'], axis=1, inplace=True)
X_train.drop(['Age Bins'], axis=1, inplace=True)
X_train.drop(['Age'], axis=1, inplace=True)

In [99]:
X_test.head()

G  GS  Cmp  Att  Yds  TD  Int  Att.1  Yds.1  Y/A  ...  Tm_SFO  Tm_TAM  \
311  11   4    1    1   16   0    0      0      0  0.0  ...       1       0   
368  16   0    0    0    0   0    0      0      0  0.0  ...       0       0   
439  16   4    0    0    0   0    0      6     24  4.0  ...       0       0   
24   16  14    0    0    0   0    0      0      0  0.0  ...       0       0   
364  15   4    0    0    0   0    0      0      0  0.0  ...       0       0   

     Tm_TEN  Tm_WAS  FantPos_QB  FantPos_RB  FantPos_TE  FantPos_WR  age_B  \
311       0       0           0           0           0           1      0   
368       0       0           0           1           0           0      0   
439       0       0           0           1           0           0      1   
24        0       0           0           0           0           1      0   
364       0       0           0           0           1           0      0   

     age_C  
311      0  
368      0  
439      0  
24       0  
364      1  

[5 rows x 61 columns]

In [100]:
X_train.head()

G  GS  Cmp  Att  Yds  TD  Int  Att.1  Yds.1   Y/A  ...  Tm_SFO  Tm_TAM  \
110  15  10    0    0    0   0    0    198    680  3.43  ...       0       0   
246  16  11    0    0    0   0    0      0      0  0.00  ...       0       0   
294  12   0    0    0    0   0    0     36    191  5.31  ...       0       0   
99   16  16    0    0    0   0    0    234    871  3.72  ...       0       1   
255  16  14    0    0    0   0    0      8     30  3.75  ...       1       0   

     Tm_TEN  Tm_WAS  FantPos_QB  FantPos_RB  FantPos_TE  FantPos_WR  age_B  \
110       0       0           0           1           0           0      0   
246       0       0           0           0           1           0      1   
294       0       0           0           1           0           0      0   
99        0       0           0           1           0           0      0   
255       0       0           0           1           0           0      1   

     age_C  
110      1  
246      0  
294      0  
99       0  
255      0  

[5 rows x 61 columns]

## Normalizing Data

In [101]:
ss = StandardScaler()
ss.fit(X_train)
X_train_scaled = ss.transform(X_train)
X_test_scaled = ss.transform(X_test)
y_train.value_counts(normalize=True)

0.0      0.062865
2.2      0.004386
1.8      0.004386
9.2      0.003655
3.1      0.003655
           ...   
10.6     0.000731
143.3    0.000731
72.7     0.000731
380.8    0.000731
116.5    0.000731
Name: PPRtarget, Length: 923, dtype: float64

## Dummy Model

In [102]:
dummy = DummyRegressor(strategy='mean')

In [103]:
dummy.fit(X_train_scaled, y_train)

DummyRegressor()

In [104]:
print(dummy.predict(y)[:10])

[89.23559942 89.23559942 89.23559942 89.23559942 89.23559942 89.23559942
 89.23559942 89.23559942 89.23559942 89.23559942]


Our Dummy model simple predicts the average fantasy point score of 89.2 for each player

## Linear Regression Simple Model

In [189]:
lr = LinearRegression()
lr.fit(X_train_scaled, y_train)

LinearRegression()

In [190]:
lr.score(X_train_scaled,y_train)

0.5631374707620538

In [107]:
lr_cv_results = cross_validate(lr, X_train_scaled, y_train, cv=5)

In [108]:
lr_cv_results

{'fit_time': array([0.00299239, 0.00199461, 0.00199366, 0.00199437, 0.00099754]),
 'score_time': array([0.0009973 , 0.00099802, 0.00099945, 0.        , 0.0009973 ]),
 'test_score': array([0.42540114, 0.58586247, 0.49845161, 0.44972751, 0.52788673])}

In [109]:
lr_test_scores = lr_cv_results['test_score']

In [110]:
lr_score = lr_test_scores.mean()
lr_score

0.49746588963594257

## Decision Tree Regressor

In [191]:
dt = DecisionTreeRegressor() 

dt.fit(X_train_scaled,y_train)

DecisionTreeRegressor()

In [112]:
dt.score(X_train_scaled,y_train)

1.0

In [137]:
dt_cv_results = cross_validate(dt, X_train_scaled, y_train, cv=5)

In [138]:
dt_cv_results

{'fit_time': array([0.01296663, 0.01296496, 0.01296544, 0.0129652 , 0.01196837]),
 'score_time': array([0.00099659, 0.00099778, 0.00099683, 0.        , 0.00099802]),
 'test_score': array([-0.16430972,  0.14662501,  0.18028091,  0.12486771,  0.11644921])}

In [139]:
dt_test_scores = dt_cv_results['test_score']
dt_score = dt_test_scores.mean()
dt_score

0.08078262499912581

## KNN Regressor

In [192]:
knn = KNeighborsRegressor(n_neighbors=12)
knn.fit(X_train_scaled, y_train)



KNeighborsRegressor(n_neighbors=12)

In [185]:
knn.score(X_train_scaled,y_train)


0.49555226897634885

In [118]:
knn_cv_results = cross_validate(knn, X_train_scaled, y_train, cv=5)
knn_cv_results

{'fit_time': array([0.00997448, 0.00997353, 0.00897598, 0.00897598, 0.00797796]),
 'score_time': array([0.02592969, 0.02393603, 0.02293849, 0.02293968, 0.02293897]),
 'test_score': array([0.30913337, 0.46233068, 0.41169172, 0.33612563, 0.34495354])}

In [119]:
knn_test_scores = knn_cv_results['test_score']
knn_score = knn_test_scores.mean()
knn_score

0.3728469888053502

## Random Forest Regressor

In [193]:
rf = RandomForestRegressor()
rf.fit(X_train_scaled, y_train)
 

RandomForestRegressor()

In [141]:
rf.score(X_train_scaled,y_train)

0.9302197608807329

In [150]:
rf_cv_results = cross_validate(rf, X_train_scaled, y_train, cv=20)
rf_cv_results

{'fit_time': array([0.89845228, 0.90422273, 0.91034389, 0.90458059, 0.89785385,
        0.88899207, 0.88326693, 0.90196347, 0.89376235, 0.88916945,
        0.89537358, 0.89244199, 0.88555479, 0.89011717, 0.89360857,
        0.9662919 , 0.92616701, 0.90261126, 0.8899982 , 0.88849354]),
 'score_time': array([0.00698137, 0.00598788, 0.00596166, 0.00698113, 0.00598311,
        0.00622368, 0.00698137, 0.00598478, 0.00615001, 0.00698233,
        0.00651288, 0.00698042, 0.00698113, 0.00698137, 0.00698185,
        0.00797892, 0.00696754, 0.00695658, 0.00599504, 0.00599194]),
 'test_score': array([0.20890779, 0.61967472, 0.33831078, 0.30717306, 0.47794902,
        0.48600731, 0.63784991, 0.64847312, 0.54281614, 0.37104018,
        0.47666742, 0.53378944, 0.57311498, 0.36138243, 0.26350195,
        0.6220372 , 0.47968742, 0.25085569, 0.67598923, 0.5465324 ])}

In [149]:
rf_test_scores = rf_cv_results['test_score']
rf_score = rf_test_scores.mean()
rf_score

0.48392947558458876

## Gradient Boost

In [194]:
gbr = GradientBoostingRegressor()
gbr.fit(X_train_scaled, y_train)

GradientBoostingRegressor()

In [125]:
gbr.score(X_train_scaled,y_train)

0.7176379817666825

In [126]:
gbr_cv_results = cross_validate(gbr, X_train_scaled, y_train, cv=3)
gbr_cv_results

{'fit_time': array([0.19207001, 0.18762922, 0.18793082]),
 'score_time': array([0.00099111, 0.00099754, 0.00099707]),
 'test_score': array([0.44371211, 0.55374651, 0.52157388])}

In [127]:
gbr_test_scores = gbr_cv_results['test_score']
gbr_score = gbr_test_scores.mean()
gbr_score

0.5063441679950428

## Choosing Best Untuned Model

In [128]:
pd.DataFrame({'Linear Regression':[lr_score],
              'KNN Regression': [knn_score],
              'Random Forest Regressor': [rf_score],
               'Gradient Boost Regressor': [gbr_score]
             })

Linear Regression  KNN Regression  Random Forest Regressor  \
0           0.497466        0.372847                 0.491486   

   Gradient Boost Regressor  
0                  0.506344

It appears that the linear Regression model is the most accurate with untuned parameters, followed by RF and Gradient Boost. 

## Applying GridSearch

## Lasso(L1) Regression

In order to apply a grid search to a linear regression model, I am going to use L1 regulization, or Lasso. Lasso is an effective technique with our data since it is a method that helps reduce the dimensionality of our data. Since we have a lot of statisical data that is correlated, Lasso regression will allow us to deal with multicollinearity. 

In [129]:
lasso_parameters = {'alpha':[0.02, 0.024, 0.025, 0.026, 0.03]}

In [130]:
grid_lr = GridSearchCV(estimator=Lasso(), param_grid = lasso_parameters, cv = 10, n_jobs=-1)

In [ ]:
grid_lr.fit(X_train_scaled, y_train)

Our first Grid search did not converge with Lasso, so lets change the tolerance

In [ ]:
lasso_parameters = {'alpha':[0.02, 0.024, 0.025, 0.026, 0.03],
                    'normalize': [True, False],
                     'fit_intercept':[True, False],
                     'selection':["cyclic","random"],
                    'tol':[0.0001, 0.001,0.01,0.1,0.5]
                   }

In [ ]:
grid_lr = GridSearchCV(estimator=Lasso(tol=1e-2), param_grid = lasso_parameters, cv = 10, n_jobs=-1)

In [ ]:
grid_lr.fit(X_train_scaled, y_train)

In [ ]:
print("\n Best parameters:\n",grid_lr.best_params_)
print("\n With a score of:\n",grid_lr.best_score_)

We seem to have dealt with the convergence issue, now lets adjust the alpha parameters since the best paramter is at the upper range, and give it more tolerance options near .01

In [ ]:
lasso_parameters = {'alpha':[5, 0.5, 0.05, 0.005, 0.0005, 1, 0.1, 0.01, 0.001, 0.0001, 0 ],
                    'normalize': [True, False],
                     'fit_intercept':[True, False],
                     'selection':["cyclic","random"],
                    'tol':[0.07,0.01,0.012,0.015]
                   }



In [ ]:
grid_lr = GridSearchCV(estimator=Lasso(), param_grid = lasso_parameters, cv = 10, n_jobs=-1)

In [ ]:
grid_lr.fit(X_train_scaled, y_train)

In [ ]:
print("\n Best parameters:\n",grid_lr.best_params_)
print("\n With a score of:\n",grid_lr.best_score_)

lets repeat the process with a range of alpha close to the best parameter of 1, and tolerance tuned again

In [ ]:
lasso_parameters = {'alpha':[.7,.9,.95, 1,1.25,1.5,2,3],
                    'normalize': [True, False],
                     'fit_intercept':[True, False],
                     'selection':["cyclic","random"],
                    'tol':[0.01,0.011,0.012,0.015]
                   }

grid_lr = GridSearchCV(estimator=Lasso(), param_grid = lasso_parameters, cv = 10, n_jobs=-1)
grid_lr.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_lr.best_params_)
print("\n With a score of:\n",grid_lr.best_score_)

Repeating one final time with a range of alpha around 1.5, leaning closer to 1

In [180]:
lasso_parameters = {'alpha':[1.25,1.3,1.33,1.35,1.38,1.4,1.42,1.45,1.48,1.5,1.55,1.6,1.65,1.7,1.75,1.8],
                    'normalize': [True, False],
                     'fit_intercept':[True, False],
                     'selection':["cyclic","random"],
                    'tol':[0.09,0.095,0.01,0.011,0.012,0.013,0.014]
                   }

grid_lr = GridSearchCV(estimator=Lasso(), param_grid = lasso_parameters, cv = 10, n_jobs=-1)
grid_lr.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_lr.best_params_)
print("\n With a score of:\n",grid_lr.best_score_)


 Best parameters:
 {'alpha': 1.6, 'fit_intercept': True, 'normalize': False, 'selection': 'random', 'tol': 0.01}

 With a score of:
 0.5169847139172541


The final accuracy of our Lasso model is 0.5168

# Random Forest Grid Search

Setting up our parameters for Random Forest, we initially give it a fairly wide spectrum of options

In [131]:
rf_parameters = { "n_estimators" : [50,100,200],
                    "max_features"   : ["auto", "sqrt", "log2"],
                 "min_samples_split" : [2,3,7],
                          "bootstrap": [True, False],
                     "max_depth": [3,10,50],
                 'min_samples_leaf': [1, 3, 6],
            }



In [132]:
grid_rf = GridSearchCV(estimator=rf, param_grid = rf_parameters, cv = 5, n_jobs=-1)

In [133]:
grid_rf.fit(X_train_scaled, y_train)

GridSearchCV(cv=5, estimator=RandomForestRegressor(), n_jobs=-1,
             param_grid={'bootstrap': [True, False], 'max_depth': [3, 10, 50],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'min_samples_leaf': [1, 3, 6],
                         'min_samples_split': [2, 3, 7],
                         'n_estimators': [50, 100, 200]})

In [134]:
print("\n Best parameters:\n",grid_rf.best_params_)
print("\n With a score of:\n",grid_rf.best_score_)


 Best parameters:
 {'bootstrap': True, 'max_depth': 3, 'max_features': 'auto', 'min_samples_leaf': 6, 'min_samples_split': 7, 'n_estimators': 100}

 With a score of:
 0.5042371492628068


Based on our initial grid search, we adjust parameters and change the cross validation back to 10

In [135]:
rf_parameters = { "n_estimators" : [75,100,150,200],
                    "max_features"   : ["auto", "sqrt", "log2"],
                 "min_samples_split" : [5,7,9],
                          "bootstrap": [True, False],
                     "max_depth": [2,3,5],
                 'min_samples_leaf': [5, 6, 8],
            }
grid_rf = GridSearchCV(estimator=rf, param_grid = rf_parameters, cv = 10, n_jobs=-1)
grid_rf.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_rf.best_params_)
print("\n With a score of:\n",grid_rf.best_score_)


 Best parameters:
 {'bootstrap': True, 'max_depth': 5, 'max_features': 'auto', 'min_samples_leaf': 8, 'min_samples_split': 9, 'n_estimators': 150}

 With a score of:
 0.5062648856631005


Updating most of our parameters but leaving n_estimators at 150 since it as remained constant:

In [151]:
rf_parameters = { "n_estimators" : [150],
                    "max_features"   : ["auto", "sqrt", "log2"],
                 "min_samples_split" : [12,14,20],
                          "bootstrap": [True, False],
                     "max_depth": [6,8,10],
                 'min_samples_leaf': [12,15,20],
            }
grid_rf = GridSearchCV(estimator=rf, param_grid = rf_parameters, cv = 10, n_jobs=-1)
grid_rf.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_rf.best_params_)
print("\n With a score of:\n",grid_rf.best_score_)


 Best parameters:
 {'bootstrap': True, 'max_depth': 6, 'max_features': 'auto', 'min_samples_leaf': 15, 'min_samples_split': 12, 'n_estimators': 150}

 With a score of:
 0.5115656941556638


Repeating again, but with the parameters very finely tuned and holding bootstrap true

In [152]:
rf_parameters = { "n_estimators" : [145,150,155],
                    "max_features"   : ["auto", "sqrt", "log2"],
                 "min_samples_split" : [10,11,12],
                          "bootstrap": [True],
                     "max_depth": [5,6,7],
                 'min_samples_leaf': [14,15,16],
            }
grid_rf = GridSearchCV(estimator=rf, param_grid = rf_parameters, cv = 10, n_jobs=-1)
grid_rf.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_rf.best_params_)
print("\n With a score of:\n",grid_rf.best_score_)


 Best parameters:
 {'bootstrap': True, 'max_depth': 7, 'max_features': 'auto', 'min_samples_leaf': 14, 'min_samples_split': 12, 'n_estimators': 150}

 With a score of:
 0.5120266710350109


In [ ]:
Repeating again

In [153]:
rf_parameters = { "n_estimators" : [148,150,152],
                    "max_features"   : ["auto", "sqrt", "log2"],
                 "min_samples_split" : [12,13,14],
                          "bootstrap": [True],
                     "max_depth": [7,8,9],
                 'min_samples_leaf': [13,14],
            }
grid_rf = GridSearchCV(estimator=rf, param_grid = rf_parameters, cv = 10, n_jobs=-1)
grid_rf.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_rf.best_params_)
print("\n With a score of:\n",grid_rf.best_score_)


 Best parameters:
 {'bootstrap': True, 'max_depth': 8, 'max_features': 'auto', 'min_samples_leaf': 14, 'min_samples_split': 12, 'n_estimators': 152}

 With a score of:
 0.5121017545667774


In [155]:
rf_parameters = { "n_estimators" : [151,152,153,154],
                    "max_features"   : ["auto", "sqrt", "log2"],
                 "min_samples_split" : [11,12],
                          "bootstrap": [True],
                     "max_depth": [8],
                 'min_samples_leaf': [14],
            }
grid_rf = GridSearchCV(estimator=rf, param_grid = rf_parameters, cv = 10, n_jobs=-1)
grid_rf.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_rf.best_params_)
print("\n With a score of:\n",grid_rf.best_score_)


 Best parameters:
 {'bootstrap': True, 'max_depth': 8, 'max_features': 'auto', 'min_samples_leaf': 14, 'min_samples_split': 12, 'n_estimators': 153}

 With a score of:
 0.5109432299977358


We have finally narrowed doing the rf parameters, and we will use this iteration's Best parameters

## Gradient Boost Grid Search

Again, giving our model a wide range of initial parameters

In [156]:
gbr_parameters = {'learning_rate': [0.0001,0.001,0.01,0.1],
                  'subsample'    : [0.3,0.5,0.8],
                  'n_estimators' : [50,200,500],
                  'max_depth'    : [3,5,8]
                  }

In [157]:
grid_gbr = GridSearchCV(estimator=gbr, param_grid = gbr_parameters, cv = 10, n_jobs=-1)

In [158]:
grid_gbr.fit(X_train_scaled, y_train)

GridSearchCV(cv=10, estimator=GradientBoostingRegressor(), n_jobs=-1,
             param_grid={'learning_rate': [0.0001, 0.001, 0.01, 0.1],
                         'max_depth': [3, 5, 8], 'n_estimators': [50, 200, 500],
                         'subsample': [0.3, 0.5, 0.8]})

In [159]:
print("\n Best parameters:\n",grid_gbr.best_params_)
print("\n With a score of:\n",grid_gbr.best_score_)


 Best parameters:
 {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 500, 'subsample': 0.3}

 With a score of:
 0.5071630937880969


Itterating through, but updating parameters around learning rate and n_estimators especially

In [161]:
gbr_parameters = {'learning_rate': [.07,0.01,0.15,.02],
                  'subsample'    : [0.2,0.3,0.5],
                  'n_estimators' : [400,500,700],
                  'max_depth'    : [2,3,4]
                  }
grid_gbr = GridSearchCV(estimator=gbr, param_grid = gbr_parameters, cv = 10, n_jobs=-1)
grid_gbr.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_gbr.best_params_)
print("\n With a score of:\n",grid_gbr.best_score_)


 Best parameters:
 {'learning_rate': 0.01, 'max_depth': 4, 'n_estimators': 400, 'subsample': 0.2}

 With a score of:
 0.5093319935192678


Updating and itterating again

In [163]:
gbr_parameters = {'learning_rate': [.09,0.01,0.11],
                  'subsample'    : [0.15,0.2,0.22],
                  'n_estimators' : [350,400,450],
                  'max_depth'    : [3,4,5]
                  }
grid_gbr = GridSearchCV(estimator=gbr, param_grid = gbr_parameters, cv = 10, n_jobs=-1)
grid_gbr.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_gbr.best_params_)
print("\n With a score of:\n",grid_gbr.best_score_)


 Best parameters:
 {'learning_rate': 0.01, 'max_depth': 4, 'n_estimators': 450, 'subsample': 0.15}

 With a score of:
 0.5108698955036108


Itterating again(multiple times) focusing on subsample and n_estimator final values

In [174]:
gbr_parameters = {'learning_rate': [0.01],
                  'subsample'    : [0.15,0.16,0.17,0.18,0.19,0.2],
                  'n_estimators' : [415,420,425,430],
                  'max_depth'    : [4]
                  }
grid_gbr = GridSearchCV(estimator=gbr, param_grid = gbr_parameters, cv = 10, n_jobs=-1)
grid_gbr.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_gbr.best_params_)
print("\n With a score of:\n",grid_gbr.best_score_)


 Best parameters:
 {'learning_rate': 0.01, 'max_depth': 4, 'n_estimators': 425, 'subsample': 0.16}

 With a score of:
 0.5100002910808819


We will use this iteration's Best parameters

## K Nearest Neighbors Grid Search

In [175]:
knn_parameters = {
    'n_neighbors': (1,5,15),
    'leaf_size': (2,3,5,10),
    'weights': ('uniform', 'distance'),
    'metric': ('minkowski', 'chebyshev')
}

In [176]:
grid_knn = GridSearchCV(estimator=knn, param_grid = knn_parameters, cv = 10, n_jobs=-1)

In [177]:
grid_knn.fit(X_train_scaled, y_train)

GridSearchCV(cv=10, estimator=KNeighborsRegressor(n_neighbors=12), n_jobs=-1,
             param_grid={'leaf_size': (2, 3, 5, 10),
                         'metric': ('minkowski', 'chebyshev'),
                         'n_neighbors': (1, 5, 15),
                         'weights': ('uniform', 'distance')})

In [178]:
print("\n Best parameters:\n",grid_knn.best_params_)
print("\n With a score of:\n",grid_knn.best_score_)


 Best parameters:
 {'leaf_size': 2, 'metric': 'minkowski', 'n_neighbors': 15, 'weights': 'distance'}

 With a score of:
 0.39038412301973613


Lets iterate through again and adjust parameters

In [183]:
knn_parameters = {
    'n_neighbors': (12,15,25,50),
    'leaf_size': (2,3,4),
    'weights': ('uniform', 'distance'),
    'metric': ('minkowski', 'chebyshev')
}
grid_knn = GridSearchCV(estimator=knn, param_grid = knn_parameters, cv = 10, n_jobs=-1)
grid_knn.fit(X_train_scaled, y_train)
print("\n Best parameters:\n",grid_knn.best_params_)
print("\n With a score of:\n",grid_knn.best_score_)


 Best parameters:
 {'leaf_size': 2, 'metric': 'minkowski', 'n_neighbors': 12, 'weights': 'distance'}

 With a score of:
 0.3914073384058933


at this point since our KNN model is scoring so much lower then our other models after intial itterations we are ending, and choosing a different model

In [181]:
pd.DataFrame({'Linear Regression':[grid_lr.best_score_],
              'KNN Regression': [grid_knn.best_score_],
              'Random Forest Regressor': [grid_rf.best_score_],
               'Gradient Boost Regressor': [grid_gbr.best_score_]
             })

Linear Regression  KNN Regression  Random Forest Regressor  \
0           0.516985        0.390384                 0.510943   

   Gradient Boost Regressor  
0                      0.51

Based on our scores, our best performing model is the Lasso Linear Regression model and we will be selecting it as our model going forward

## Evaluating vs Test Data

Our final Lasso model had its best performance with the parameters of:

 {'alpha': 1.6, 'fit_intercept': True, 'normalize': False, 'selection': 'random', 'tol': 0.01}

In [200]:
lasso=Lasso(alpha= 1.6, fit_intercept= True, normalize=False, selection= 'random', tol= 0.01)
lasso.fit(X_train_scaled,y_train)

Lasso(alpha=1.6, selection='random', tol=0.01)

In [201]:
lasso.score(X_train_scaled,y_train)

0.5458169551258258

In [202]:
lasso.score(X_test_scaled,y_test)

0.5226154976144299

Or final model has an r-squared score of 0.546 on the training data, and 0.52 on the test data. In other words, just over half of the variance in a Fantasy Football players scoring in a season can be explained by our model. Although initially this doesnt seem great, there are a number of high variance events that are very difficult to explain. For example, a healthy player may score 250 fantasy points in a season, and then due to a nagging hamstring inury that we have no data on, may drop to half of that the next season. Injury is the largest introducer of variance in our system that is unadressed by the model, but there are many others, such as coaching staff changes, contract negotiations, off-field behavorial issues, etc. 

## Visualizing Model Predictions

First, Checking to make sure our prediccted PPR points match our data frame:

In [238]:
X_test_scaled.shape

(343, 61)

In [239]:
Final_predictions = lasso.predict(X_test_scaled)
Final_predictions.shape

(343,)

In [240]:
X_test.shape

(343, 62)

In [241]:
X_test.head()

G  GS  Cmp  Att  Yds  TD  Int  Att.1  Yds.1  Y/A  ...  Tm_TAM  Tm_TEN  \
311  11   4    1    1   16   0    0      0      0  0.0  ...       0       0   
368  16   0    0    0    0   0    0      0      0  0.0  ...       0       0   
439  16   4    0    0    0   0    0      6     24  4.0  ...       0       0   
24   16  14    0    0    0   0    0      0      0  0.0  ...       0       0   
364  15   4    0    0    0   0    0      0      0  0.0  ...       0       0   

     Tm_WAS  FantPos_QB  FantPos_RB  FantPos_TE  FantPos_WR  age_B  age_C  \
311       0           0           0           0           1      0      0   
368       0           0           1           0           0      0      0   
439       0           0           1           0           0      1      0   
24        0           0           0           0           1      0      0   
364       0           0           0           1           0      0      1   

     PredictedPPR  
311     48.739814  
368     28.887535  
439     28.152624  
24     206.335426  
364     32.455555  

[5 rows x 62 columns]

In [245]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 343 entries, 311 to 382
Data columns (total 62 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   G             343 non-null    int64  
 1   GS            343 non-null    int64  
 2   Cmp           343 non-null    int64  
 3   Att           343 non-null    int64  
 4   Yds           343 non-null    int64  
 5   TD            343 non-null    int64  
 6   Int           343 non-null    int64  
 7   Att.1         343 non-null    int64  
 8   Yds.1         343 non-null    int64  
 9   Y/A           343 non-null    float64
 10  TD.1          343 non-null    int64  
 11  Tgt           343 non-null    int64  
 12  Rec           343 non-null    int64  
 13  Yds.2         343 non-null    int64  
 14  Y/R           343 non-null    float64
 15  TD.2          343 non-null    int64  
 16  Fmb           343 non-null    float64
 17  FL            343 non-null    int64  
 18  TD.3          343 non-null  

In [242]:
X_test['PredictedPPR']=Final_predictions.tolist()
X_test.head()

G  GS  Cmp  Att  Yds  TD  Int  Att.1  Yds.1  Y/A  ...  Tm_TAM  Tm_TEN  \
311  11   4    1    1   16   0    0      0      0  0.0  ...       0       0   
368  16   0    0    0    0   0    0      0      0  0.0  ...       0       0   
439  16   4    0    0    0   0    0      6     24  4.0  ...       0       0   
24   16  14    0    0    0   0    0      0      0  0.0  ...       0       0   
364  15   4    0    0    0   0    0      0      0  0.0  ...       0       0   

     Tm_WAS  FantPos_QB  FantPos_RB  FantPos_TE  FantPos_WR  age_B  age_C  \
311       0           0           0           0           1      0      0   
368       0           0           1           0           0      0      0   
439       0           0           1           0           0      1      0   
24        0           0           0           0           1      0      0   
364       0           0           0           1           0      0      1   

     PredictedPPR  
311     48.739814  
368     28.887535  
439     28.152624  
24     206.335426  
364     32.455555  

[5 rows x 62 columns]

In [243]:
Final_df = X_test.sort_values(by=['PredictedPPR'], ascending=False)

In [244]:
Final_df.head()

G  GS  Cmp  Att   Yds  TD  Int  Att.1  Yds.1   Y/A  ...  Tm_TAM  Tm_TEN  \
20  16  16  422  608  4924  35    7     33    125  3.79  ...       0       0   
44  16  16  364  561  4688  32   12     43    108  2.51  ...       0       0   
5   16  16    0    0     0   0    0     22    151  6.86  ...       0       0   
13  16  16  341  516  4110  31    5     75    342  4.56  ...       0       0   
9   15  15    0    0     0   0    0      0      0  0.00  ...       0       0   

    Tm_WAS  FantPos_QB  FantPos_RB  FantPos_TE  FantPos_WR  age_B  age_C  \
20       0           1           0           0           0      0      1   
44       0           1           0           0           0      0      0   
5        0           0           0           0           1      0      0   
13       0           1           0           0           0      0      1   
9        0           0           0           0           1      1      0   

    PredictedPPR  
20    288.277788  
44    276.655971  
5     276.011842  
13    266.493735  
9     264.972578  

[5 rows x 62 columns]

From this we can see that the highest predicted NFL player's PPR prediction for the next year, from our test set, if we reference back, is 2018 Falcon's QB Matt Ryan with 288.3. How Accurate is this? he followed up the following year with a score of 267.3, so not too far off!

However, this is just one datapoint, lets take a look at how the model performed overall. 